ls1intum/Artemis
# Feature/programming exercises management/update selected exercises

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de and https://artemistest2.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html), **except for one point because I had to create a new component.**

### Motivation and Context


This PR has been created because of a suggestion from an instructor using ArTEMiS in his/her seminar, stating 
that having an exercise sheet each week with multiple exercises, required quite some time to manage/update.

Examples reported: 
- If the deadline changes, the deadline must be changed for all three exercises rather than once.
- Currently, if we want to download the repository and results, we have to do so for each exercise.

Therefore, the possibility to update or download repositories of multiple exercises together could save some time in this context.

### Description

An additional checkbox has been added in the programming exercises list. Additional checks ensure that this will be hidden for users without instructor permissions within the course.
As soon as at least one programming-exercise has been selected, two additional buttons will appear at the bottom of the list, which if pressed will open their respective modal component ("Edit selected" and "Download Repositories").

The "Edit selected" modal corresponds to a new modal component, which will allow to edit the timeline of the exercises.
For the "Download Repositories" modal, the currently existing one has been adapted and reused. (`programming-assessment-repo-export-dialog.component.ts`)

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Choose a Course where the user has instructor permissions
4. Select multiple programming exercises using the checkbox
5. Click the "Edit Selected" button and set a new Release or Due date
6. Save
7. Check if the selected exercises have now been updated with the new dates
8. Select programming exercises that have at least one participation
9. Click the "Download repositories" and export them
10. Choose a different Course where the user does not have instructor permissions
11. No checkboxes should be displayed

### Test Coverage



programming-exercise-edit-selected.component.ts: 69,64%
programming-exercise.component.ts: 70,37%
programming-assessment-repo-export-dialog.component.ts: 28.07%



### Screenshots


![Screenshot from 2020-11-24 09-39-06](https://user-images.githubusercontent.com/45761921/100069422-08c3df00-2e39-11eb-8694-0ed967f16580.png)
![Screenshot from 2020-11-24 09-39-19](https://user-images.githubusercontent.com/45761921/100069426-095c7580-2e39-11eb-9a57-beb6e395ef9d.png)
![Screenshot from 2020-11-24 09-39-28](https://user-images.githubusercontent.com/45761921/100069428-09f50c00-2e39-11eb-8724-d1f7d748491b.png)
